src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -134,5 +127,329 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+
+            BoundInterpolatedString constructWithData(ImmutableArray<BoundExpression> parts, InterpolatedStringBuilderData? data)
+                => new BoundInterpolatedString(
+                    unconvertedInterpolatedString.Syntax,
+                    data,
+                    parts,
+                    unconvertedInterpolatedString.ConstantValue,
+                    unconvertedInterpolatedString.Type,
+                    unconvertedInterpolatedString.HasErrors);
+
+            bool tryBindAsBuilderType([NotNullWhen(true)] out BoundInterpolatedString? result)
+            {
+                result = null;
+                if (unconvertedInterpolatedString.Parts.ContainsAwaitExpression())
+                {
+                    // PROTOTYPE(interp-string): For interpolated strings used as strings, we could evaluate components eagerly
+                    // and always use the builder.
+                    return false;
+                }
+
+                var interpolatedStringBuilderType = Compilation.GetWellKnownType(WellKnownType.System_Runtime_CompilerServices_InterpolatedStringBuilder);
+                if (interpolatedStringBuilderType is MissingMetadataTypeSymbol)
+                {
+                    return false;
+                }
+
+                diagnostics.Add(interpolatedStringBuilderType.GetUseSiteInfo(), unconvertedInterpolatedString.Syntax.Location);
+
+                // We satisfy the conditions for using an interpolated string builder. Bind all the builder calls unconditionally, so that if
+                // there are errors we get better diagnostics than "could not convert to object."
+                var (appendCalls, usesBoolReturn) = BindInterpolatedStringAppendCalls(unconvertedInterpolatedString, interpolatedStringBuilderType, diagnostics);
+
+                // Prior to C# 10, all types in an interpolated string expression needed to be convertible to `object`. After 10, some types
+                // (such as Span<T>) that are not convertible to `object` are permissible as interpolated string components, provided there
+                // is an applicable AppendFormatted method that accepts them. To preserve langversion, we therefore make sure all components
+                // are convertible to object if the current langversion is lower than the interpolation feature
+                TypeSymbol? objectType = null;
+                BindingDiagnosticBag? conversionDiagnostics = null;
+                var needToCheckConversionToObject = !Compilation.IsFeatureEnabled(MessageID.IDS_FeatureImprovedInterpolatedStrings) && diagnostics.AccumulatesDiagnostics;
+                if (needToCheckConversionToObject)
+                {
+                    objectType = GetSpecialType(SpecialType.System_Object, diagnostics, unconvertedInterpolatedString.Syntax);
+                    conversionDiagnostics = BindingDiagnosticBag.GetInstance(template: diagnostics);
```

Comment:
>template: diagnostics [](http://example.com/codeflow?start=77&length=21)

It looks like we don't need to track dependencies after all, regardless of what the ```diagnosics``` is doing.